### ISSUE TITLE
Error: No matching export in "xxx" for import "default"

### ISSUE THREAD

**Issue by datWeazel** [hyrious/esbuild-plugin-commonjs#5]
Using the plugin results in the following error when trying to build.
A similar error happens when I try to use typeorm in my project. The error that occurs if I use typeorm is with "uuid" not having a default import.)

```
✘ [ERROR] No matching export in "node_modules/acorn-loose/dist/acorn-loose.mjs" for import "default"

    node_modules/@mikro-orm/core/utils/Utils.js:1:251:
      1 │ ...ort __import_os from "os";import __import_url from "url";import __import_fsExtra from "fs-extra";import __import_crypto from "crypto";import __import_acornLoose from "acorn-loose";import __import_acornWalk from "acorn-walk";import __import___clone from "./clone";import __import____enums from "../enu...
```

**Comment by datWeazel** [comment]
This is the error I get if I try using typeorm: 
```
✘ [ERROR] No matching export in "node_modules/uuid/dist/esm-node/index.js" for import "default"

    node_modules/typeorm/query-builder/InsertQueryBuilder.js:1:7:
      1 │ import __import_uuid from "uuid";import __import____driver_DriverUtils from "../driver/DriverUtils";import __import____error from "../error";import __import____error_InsertValuesMissingError from "../error/InsertValuesMissingError";import __import____error_ReturningStatementNotSupportedError from "../e...
        ╵        ~~~~~~~~~~~~~


```

**Comment by hyrious** [comment]
This is because some library (like `uuid` or `acorn-loose`) has exported both CommonJS and ES module files, where importing them needs different style of codes:

```js
// original code
require("uuid")

// possible replacement in esm context
import __import_uuid from "uuid"

// another possible replacement
import * as __import_uuid from "uuid"
```

The 2 replacement above has different semantics depending on how the `uuid` module look like. Even when the `uuid` is a cjs module the behavior will differ when esbuild detects Node.js ESM context (i.e. file has '.mjs' extension or there be "type": "module" in package.json).

So it is not always safe to choose which style to convert. This plugin assumes a common situation where you're bundling a CommonJS module into ESM format, and the dependencies of that module are also in CommonJS, like how `react-dom` depends on `react`.

```
-->: depends on

Your esm code --> react-dom (cjs) --> react (cjs) ✅

Your esm code --> typeorm (cjs) --> uuid (esm) ❌
```

So maybe I should add an option to let you opt-in the style of importing external modules, like [requireReturnsDefault](https://www.npmjs.com/package/@rollup/plugin-commonjs#requirereturnsdefault)

```js
requireReturnsDefault: (mod) => !['uuid'].includes(mod)
```

**Comment by hyrious** [comment]
Just published @hyrious/esbuild-plugin-commonjs@0.2.3 with the new option `requireReturnsDefault`, maybe you can give it a try.

**Comment by datWeazel** [comment]
Little nitpick: The type declarations seem to be faulty.
`declare function commonjs({ filter, transform, transformConfig, }?: CommonJSOptions): Plugin;`
should be 
`declare function commonjs({ filter, transform, transformConfig, requireReturnsDefault, ignore, }?: CommonJSOptions): Plugin;`
shouldn't it? 

---

Anyway .. using 
`commonjs({ requireReturnsDefault: (mod) => !["uuid", "glob"].includes(mod) })`
I can now build everything. Sadly it still doesn't work when I try to run my code but that's another issue I guess.

In case you are interested: 
`[00:17:50][Error] ReferenceError: __dirname is not defined in ES module scope`
platform for esbuild is `node` and format `esm`. 

I need my code to be bundled to a ESModule to work properly in the environment I run the code. 

But as I said that is probably a different issue so thank you so far. :)

**Comment by hyrious** [comment]
> The type declarations seem to be faulty.

My mistake, just fixed that on v0.2.4.

> `ReferenceError: __dirname is not defined in ES module scope`

There's no global variable `__dirname` in ES modules context, instead they are replaced by `import.meta.url`. See https://nodejs.org/docs/latest/api/esm.html#differences-between-es-modules-and-commonjs.

Code depending on the location of itself is often not working well with bundlers, for example:

```js
fs.readFileSync(path.join(__dirname, '../asset.txt'))
```

The above example code assumes it can find asset.txt at runtime right in its _parent_ directory. However bundlers won't know this infomation. Some old versions of webpack may execute a crazy logic that it simulates the whole filesystem once matching code patterns like that.

In this case, you'd better just not bundle the dependency and leave it as-is.
